lib/manageiq/session_store.rb

```diff
@@ -0,0 +1,88 @@
+module ManageIQ
+  module SessionStore
+    def self.store
+      @store
+    end
+
+    def self.store=(session_store)
+      adapter = case session_store
+                when :active_record_store then ActiveRecordStoreAdapter.new
+                when :memory_store        then MemoryStoreAdapter.new
+                when :mem_cache_store     then MemCacheStoreAdapter.new
+                else
+                  raise "Please provide a valid Session Store"
+                end
+
+      adapter.patch_session_store_class
+      set_store_config(adapter)
+
+      @store = adapter.class_to_patch.new(nil, adapter.session_options)
+    end
+
+    def self.set_store_config(adapter)
+      Vmdb::Application.config.session_store(*adapter.session_store_config_args)
+      msg = "Using session_store: #{Vmdb::Application.config.session_store}"
+      $log.info("MIQ(SessionStore) #{msg}")
+      puts "** #{msg}" if !Rails.env.production? && adapter.type != :mem_cache_store
+    end
+
+    def self.revoke_all(session_ids)
+      store.delete_sessions(session_ids)
+    end
+
+    class AbstractStoreAdapter
+      ClassToPatch = Class.new
+
+      def self.class_to_patch
+        ClassToPatch
+      end
+
+      def self.patch_session_store_class
+        return unless needs_patch?
+        @_already_patched = true
+
+        class_to_patch.prepend self::SessionStorePatch
+      rescue => e
+        @_already_patched = false
+        raise e
+      end
+
+      def self.needs_patch?
+        @_already_patched ||= false
+
+        @_already_patched || class_to_patch != AbstractStoreAdapter.class_to_patch
```

Comment:
All of this "already patched" stuff seems like it can be removed in favor of checking the ancestry of the class_to_patch.  Something like:

```ruby
class_to_patch.ancestors.include?(self::SessionStorePatch)
```

Reply:
No longer relevant, but I did do this for a bit, and it got rid of a lot of extra code.  +1